File: internal/k8s/discovery.go
package k8s

import (
	"context"
	"fmt"
	"sync"
)

// Discovery scans Kubernetes resources for common issues
type Discovery struct {
	client *Client
}

// NewDiscovery creates a new Discovery engine
func NewDiscovery(client *Client) *Discovery {
	return &Discovery{
		client: client,
	}
}

// ScanAll scans all resources in the specified namespace for issues
// Runs multiple scans in parallel for efficiency
func (d *Discovery) ScanAll(ctx context.Context, namespace string) ([]Issue, error) {
	var wg sync.WaitGroup
	var mu sync.Mutex
	allIssues := []Issue{}

	// Channel to collect errors
	errChan := make(chan error, 5)

	// Scan pods
	wg.Add(1)
	go func() {
		defer wg.Done()
		issues, err := d.ScanPods(ctx, namespace)
		if err != nil {
			errChan <- fmt.Errorf("pod scan: %w", err)
			return
		}
		mu.Lock()
		allIssues = append(allIssues, issues...)
		mu.Unlock()
	}()

	// Scan deployments
	wg.Add(1)
	go func() {
		defer wg.Done()
		issues, err := d.ScanDeployments(ctx, namespace)
		if err != nil {
			errChan <- fmt.Errorf("deployment scan: %w", err)
			return
		}
		mu.Lock()
		allIssues = append(allIssues, issues...)
		mu.Unlock()
	}()

	// Scan events
	wg.Add(1)
	go func() {
		defer wg.Done()
		issues, err := d.ScanEvents(ctx, namespace)
		if err != nil {
			errChan <- fmt.Errorf("event scan: %w", err)
			return
		}
		mu.Lock()
		allIssues = append(allIssues, issues...)
		mu.Unlock()
	}()

	// Scan nodes (cluster-wide)
	wg.Add(1)
	go func() {
		defer wg.Done()
		issues, err := d.ScanNodes(ctx)
		if err != nil {
			errChan <- fmt.Errorf("node scan: %w", err)
			return
		}
		mu.Lock()
		allIssues = append(allIssues, issues...)
		mu.Unlock()
	}()

	// Wait for all scans to complete
	wg.Wait()
	close(errChan)

	// Collect any errors
	var scanErrors []error
	for err := range errChan {
		scanErrors = append(scanErrors, err)
	}

	// Return issues even if some scans failed
	if len(scanErrors) > 0 {
		return allIssues, fmt.Errorf("some scans failed: %v", scanErrors)
	}

	return allIssues, nil
}

// ScanPods scans pods for common issues
func (d *Discovery) ScanPods(ctx context.Context, namespace string) ([]Issue, error) {
	pods, err := d.client.GetPods(ctx, namespace)
	if err != nil {
		return nil, err
	}

	var issues []Issue

	for _, pod := range pods {
		// Check pod status
		if pod.Status != "Running" && pod.Status != "Succeeded" {
			priority := PriorityWarning
			statusHint := ""

			switch pod.Status {
			case "Failed", "CrashLoopBackOff", "Error":
				priority = PriorityCritical
				statusHint = fmt.Sprintf(" Container is crashing. Check logs with: kubectl logs %s -n %s --previous",
					pod.Name, pod.Namespace)

			case "Pending":
				statusHint = fmt.Sprintf(" Pod cannot be scheduled. Check events with: kubectl get events -n %s --field-selector involvedObject.name=%s",
					pod.Namespace, pod.Name)

			case "ImagePullBackOff", "ErrImagePull":
				priority = PriorityCritical
				statusHint = " Cannot pull container image. Check image name and registry access."

			case "OOMKilled":
				priority = PriorityCritical
				statusHint = " Pod killed due to out of memory. Increase memory limits."
			}

			issues = append(issues, Issue{
				ID:           GenerateIssueID(pod.Name, "status"),
				Title:        fmt.Sprintf("Pod %s in %s state", pod.Name, pod.Status),
				Description:  fmt.Sprintf("Pod is in %s state instead of Running.%s", pod.Status, statusHint),
				Priority:     priority,
				ResourceType: "pod",
				ResourceName: pod.Name,
				Namespace:    pod.Namespace,
			})
		}

		// Check for container-specific issues
		for _, cs := range pod.ContainerStates {
			if cs.State == "waiting" && cs.Reason != "" {
				priority := PriorityWarning

				switch cs.Reason {
				case "CrashLoopBackOff", "ImagePullBackOff", "ErrImagePull":
					priority = PriorityCritical
				}

				issues = append(issues, Issue{
					ID:          GenerateIssueID(pod.Name, fmt.Sprintf("container-%s", cs.Name)),
					Title:       fmt.Sprintf("Container %s in pod %s: %s", cs.Name, pod.Name, cs.Reason),
					Description: fmt.Sprintf("Container is waiting: %s. %s", cs.Reason, cs.Message),
					Priority:    priority,
					ResourceType: "pod",
					ResourceName: pod.Name,
					Namespace:    pod.Namespace,
				})
			}
		}

		// Check restart count
		if pod.Restarts > 5 {
			issues = append(issues, Issue{
				ID:    GenerateIssueID(pod.Name, "restarts"),
				Title: fmt.Sprintf("High restart count for %s", pod.Name),
				Description: fmt.Sprintf("Pod has restarted %d times. Check crash logs: kubectl logs %s -n %s --previous. Review events: kubectl get events -n %s --field-selector involvedObject.name=%s",
					pod.Restarts, pod.Name, pod.Namespace, pod.Namespace, pod.Name),
				Priority:     PriorityWarning,
				ResourceType: "pod",
				ResourceName: pod.Name,
				Namespace:    pod.Namespace,
			})
		}

		// Check readiness
		if !pod.Ready && pod.Status == "Running" {
			issues = append(issues, Issue{
				ID:          GenerateIssueID(pod.Name, "readiness"),
				Title:       fmt.Sprintf("Pod %s not ready", pod.Name),
				Description: fmt.Sprintf("Pod is running but readiness probe is failing. Check logs and probe configuration. Run: kubectl describe pod %s -n %s to see probe details.", pod.Name, pod.Namespace),
				Priority:    PriorityWarning,
				ResourceType: "pod",
				ResourceName: pod.Name,
				Namespace:    pod.Namespace,
			})
		}
	}

	return issues, nil
}

// ScanDeployments scans deployments for replica mismatches
func (d *Discovery) ScanDeployments(ctx context.Context, namespace string) ([]Issue, error) {
	deployments, err := d.client.GetDeployments(ctx, namespace)
	if err != nil {
		return nil, err
	}

	var issues []Issue

	for _, deployment := range deployments {
		// Check replica mismatches
		if deployment.Replicas != deployment.ReadyReplicas {
			issues = append(issues, Issue{
				ID:          GenerateIssueID(deployment.Name, "replicas"),
				Title:       fmt.Sprintf("Replica mismatch in %s", deployment.Name),
				Description: fmt.Sprintf("Expected %d replicas, %d ready", deployment.Replicas, deployment.ReadyReplicas),
				Priority:    PriorityWarning,
				ResourceType: "deployment",
				ResourceName: deployment.Name,
				Namespace:    deployment.Namespace,
			})
		}

		// Check availability
		if deployment.AvailableReplicas < deployment.Replicas {
			priority := PriorityWarning
			if deployment.AvailableReplicas == 0 {
				priority = PriorityCritical
			}

			issues = append(issues, Issue{
				ID:          GenerateIssueID(deployment.Name, "availability"),
				Title:       fmt.Sprintf("Deployment %s unavailable", deployment.Name),
				Description: fmt.Sprintf("Only %d of %d replicas available", deployment.AvailableReplicas, deployment.Replicas),
				Priority:    priority,
				ResourceType: "deployment",
				ResourceName: deployment.Name,
				Namespace:    deployment.Namespace,
			})
		}
	}

	return issues, nil
}

// ScanEvents scans for Warning and Error events
func (d *Discovery) ScanEvents(ctx context.Context, namespace string) ([]Issue, error) {
	events, err := d.client.GetEvents(ctx, namespace, false) // Only non-Normal events
	if err != nil {
		return nil, err
	}

	var issues []Issue

	// Limit to recent events (e.g., last 10)
	maxEvents := 10
	if len(events) > maxEvents {
		events = events[:maxEvents]
	}

	for _, event := range events {
		priority := PriorityInfo
		if event.Type == "Warning" {
			priority = PriorityWarning
		} else if event.Type == "Error" {
			priority = PriorityCritical
		}

		issues = append(issues, Issue{
			ID:           GenerateIssueID(event.InvolvedObject, event.Reason),
			Title:        fmt.Sprintf("%s: %s", event.InvolvedObject, event.Reason),
			Description:  event.Message,
			Priority:     priority,
			ResourceType: event.InvolvedKind,
			ResourceName: event.InvolvedObject,
			Namespace:    event.InvolvedNamespace,
		})
	}

	return issues, nil
}

// ScanNodes scans nodes for common issues
func (d *Discovery) ScanNodes(ctx context.Context) ([]Issue, error) {
	nodes, err := d.client.GetNodes(ctx)
	if err != nil {
		return nil, err
	}

	var issues []Issue

	for _, node := range nodes {
		// Check node status
		if node.Status != "Ready" {
			issues = append(issues, Issue{
				ID:           GenerateIssueID(node.Name, "status"),
				Title:        fmt.Sprintf("Node %s is %s", node.Name, node.Status),
				Description:  fmt.Sprintf("Node is not in Ready state. Check: kubectl describe node %s", node.Name),
				Priority:     PriorityCritical,
				ResourceType: "node",
				ResourceName: node.Name,
				Namespace:    "", // Nodes are cluster-scoped
			})
		}
	}

	return issues, nil
}

File: src/discovery.py
import asyncio
import re
from datetime import datetime, timezone
from typing import List, Dict, Any
try:
    from .k8s_client import K8sClient
    from .utils import Issue, Priority, generate_issue_id, get_pod_namespace
except ImportError:
    from k8s_client import K8sClient
    from utils import Issue, Priority, generate_issue_id, get_pod_namespace

class Discovery:
    def __init__(self, k8s_client: K8sClient):
        self.k8s_client = k8s_client
        self.namespace = get_pod_namespace()

    async def scan_all(self, namespace: str = None) -> List[Issue]:
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace

        # Parallel scanning for efficiency - including advanced infrastructure
        tasks = [
            self._scan_pods(ns),
            self._scan_deployments(ns),
            self._scan_events(ns),
            self._scan_service_mesh_issues(ns),
            self._scan_network_policies(ns),
            self._scan_policy_violations(ns),
        ]

        results = await asyncio.gather(*tasks, return_exceptions=True)

        for result in results:
            if isinstance(result, list):
                issues.extend(result)
            else:
                print(f"Discovery error: {result}")

        return issues

    async def _scan_pods(self, namespace: str = None) -> List[Issue]:
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            pods = await self.k8s_client.get_pods(ns)
        except Exception as e:
            print(f"Error scanning pods: {e}")
            return issues

        for pod in pods:
            # Check pod status
            if pod["status"] not in ["Running", "Succeeded"]:
                priority = Priority.CRITICAL if pod["status"] in ["Failed", "CrashLoopBackOff", "Error"] else Priority.WARNING
                status_hint = ""
                if pod["status"] in ["CrashLoopBackOff", "Error"]:
                    status_hint = f" Container is crashing. Check logs with: kubectl logs {pod['name']} -n {pod['namespace']} --previous"
                elif pod["status"] == "Pending":
                    status_hint = f" Pod cannot be scheduled. Check events with: kubectl get events -n {pod['namespace']} --field-selector involvedObject.name={pod['name']}"
                elif pod["status"] == "Failed":
                    status_hint = f" Pod has failed. Check describe output: kubectl describe pod {pod['name']} -n {pod['namespace']}"

                issues.append(Issue(
                    id=generate_issue_id(pod["name"], "status"),
                    title=f"Pod {pod['name']} in {pod['status']} state",
                    description=f"Pod is in {pod['status']} state instead of Running.{status_hint}",
                    priority=priority,
                    resource_type="pod",
                    resource_name=pod["name"],
                    namespace=pod["namespace"]
                ))

            # Check restart count
            if pod["restarts"] > 5:
                pod_name = pod["name"]
                pod_ns = pod["namespace"]
                issues.append(Issue(
                    id=generate_issue_id(pod_name, "restarts"),
                    title=f"High restart count for {pod_name}",
                    description=f"Pod has restarted {pod['restarts']} times. Check crash logs: kubectl logs {pod_name} -n {pod_ns} --previous. Review events: kubectl get events -n {pod_ns} --field-selector involvedObject.name={pod_name}",
                    priority=Priority.WARNING,
                    resource_type="pod",
                    resource_name=pod_name,
                    namespace=pod_ns
                ))

            # Check readiness
            if not pod["ready"] and pod["status"] == "Running":
                pod_name = pod["name"]
                pod_ns = pod["namespace"]
                issues.append(Issue(
                    id=generate_issue_id(pod_name, "readiness"),
                    title=f"Pod {pod_name} not ready",
                    description=f"Pod is running but readiness probe is failing. Check logs and probe configuration. Run: kubectl describe pod {pod_name} -n {pod_ns} to see probe details.",
                    priority=Priority.WARNING,
                    resource_type="pod",
                    resource_name=pod_name,
                    namespace=pod_ns
                ))

        return issues

    async def _scan_deployments(self, namespace: str = None) -> List[Issue]:
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            deployments = await self.k8s_client.get_deployments(ns)
        except Exception as e:
            print(f"Error scanning deployments: {e}")
            return issues

        for deployment in deployments:
            # Check replica mismatches
            if deployment["replicas"] != deployment["ready"]:
                issues.append(Issue(
                    id=generate_issue_id(deployment["name"], "replicas"),
                    title=f"Replica mismatch in {deployment['name']}",
                    description=f"Expected {deployment['replicas']} replicas, {deployment['ready']} ready",
                    priority=Priority.WARNING,
                    resource_type="deployment",
                    resource_name=deployment["name"],
                    namespace=deployment["namespace"]
                ))

            # Check availability
            if deployment["available"] < deployment["replicas"]:
                issues.append(Issue(
                    id=generate_issue_id(deployment["name"], "availability"),
                    title=f"Deployment {deployment['name']} unavailable",
                    description=f"Only {deployment['available']} of {deployment['replicas']} replicas available",
                    priority=Priority.CRITICAL,
                    resource_type="deployment",
                    resource_name=deployment["name"],
                    namespace=deployment["namespace"]
                ))

        return issues

    async def _scan_events(self, namespace: str = None) -> List[Issue]:
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            events = await self.k8s_client.get_events(ns)
        except Exception as e:
            print(f"Error scanning events: {e}")
            return issues

        # Group events by object and type
        event_groups = {}
        for event in events:
            key = f"{event['object_kind']}-{event['object_name']}-{event['reason']}"
            if key not in event_groups:
                event_groups[key] = []
            event_groups[key].append(event)

        for key, group in event_groups.items():
            # Focus on recent warning events
            recent_events = [e for e in group if self._is_recent(e["timestamp"])]

            if recent_events:
                latest_event = recent_events[0]  # Events are sorted by timestamp

                # Priority based on event type
                if "Error" in latest_event["reason"] or "Failed" in latest_event["reason"]:
                    priority = Priority.CRITICAL
                elif latest_event["type"] == "Warning":
                    priority = Priority.WARNING
                else:
                    priority = Priority.INFO

                # Clean up message
                message = latest_event["message"]
                if len(message) > 200:
                    message = message[:197] + "..."

                issues.append(Issue(
                    id=generate_issue_id(latest_event["object_name"], "event"),
                    title=f"{latest_event['reason']} on {latest_event['object_kind']}/{latest_event['object_name']}",
                    description=message,
                    priority=priority,
                    resource_type=latest_event["object_kind"].lower(),
                    resource_name=latest_event["object_name"],
                    namespace=latest_event["namespace"]
                ))

        return issues

    def _is_recent(self, timestamp) -> bool:
        if not timestamp:
            return False

        age = datetime.now(timezone.utc) - timestamp
        return age.total_seconds() < 3600  # Last hour

    async def get_resource_health(self, namespace: str = None) -> Dict[str, Any]:
        """Get overall cluster health summary"""
        # Support "all" or empty string for all namespaces
        if namespace == "all" or namespace == "":
            ns = None
        elif namespace is None:
            # If no namespace specified, use all namespaces by default
            ns = None
        else:
            ns = namespace

        try:
            pods = await self.k8s_client.get_pods(ns)
        except Exception as e:
            print(f"Error getting pods for health check: {e}")
            pods = []

        try:
            deployments = await self.k8s_client.get_deployments(ns)
        except Exception as e:
            print(f"Error getting deployments for health check: {e}")
            deployments = []

        try:
            events = await self.k8s_client.get_events(ns)
        except Exception as e:
            print(f"Error getting events for health check: {e}")
            events = []

        total_pods = len(pods)
        healthy_pods = len([p for p in pods if p.get("status") == "Running" and p.get("ready")])

        total_deployments = len(deployments)
        healthy_deployments = len([d for d in deployments if d.get("ready") == d.get("replicas")])

        warning_events = len([e for e in events if e.get("type") == "Warning"])
        error_events = len([e for e in events if "Error" in str(e.get("reason", "")) or "Fail" in str(e.get("reason", ""))])

        return {
            "pods": {"total": total_pods, "healthy": healthy_pods},
            "deployments": {"total": total_deployments, "healthy": healthy_deployments},
            "events": {"warnings": warning_events, "errors": error_events},
            "overall_health": "healthy" if healthy_pods == total_pods and healthy_deployments == total_deployments and total_pods > 0 else "degraded"
        }

    async def _scan_service_mesh_issues(self, namespace: str = None) -> List[Issue]:
        """Detect service mesh (Istio) related issues"""
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            pods = await self.k8s_client.get_pods(ns)

            for pod in pods:
                labels = pod.get("labels", {})
                pod_name = pod["name"]
                pod_ns = pod["namespace"]

                # Check for Istio sidecar issues
                if "istio-injection" in str(labels) or any("istio" in k.lower() for k in labels.keys()):
                    # Check if pod is running but sidecar might be failing
                    if pod["status"] == "Running" and not pod.get("ready"):
                        # Could be sidecar not ready
                        issues.append(Issue(
                            id=generate_issue_id(pod_name, "istio-sidecar"),
                            title=f"Istio sidecar issue in {pod_name}",
                            description=f"Pod appears to have Istio injection but may have sidecar issues. Check: kubectl logs {pod_name} -n {pod_ns} -c istio-proxy. Verify sidecar: kubectl get pod {pod_name} -n {pod_ns} -o jsonpath='{{.spec.containers[*].name}}'",
                            priority=Priority.WARNING,
                            resource_type="pod",
                            resource_name=pod_name,
                            namespace=pod_ns
                        ))
        except Exception as e:
            # Silently fail for service mesh scanning - not all clusters have it
            pass

        return issues

    async def _scan_network_policies(self, namespace: str = None) -> List[Issue]:
        """Detect network policy (Cilium/NetworkPolicy) related issues"""
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            # Check for pods that might be blocked by network policies
            pods = await self.k8s_client.get_pods(ns)
            events = await self.k8s_client.get_events(ns)

            # Look for network-related errors in events
            network_errors = [e for e in events if any(keyword in e.get("message", "").lower()
                for keyword in ["network", "policy", "forbidden", "denied", "cilium"])]

            for event in network_errors[-5:]:  # Last 5 network errors
                if "network" in event.get("message", "").lower() or "policy" in event.get("message", "").lower():
                    issues.append(Issue(
                        id=generate_issue_id(event["object_name"], "network-policy"),
                        title=f"Network policy issue: {event['reason']}",
                        description=f"{event['message']}. Check network policies: kubectl get networkpolicies -n {event['namespace']}. For Cilium: kubectl get cnp,ccnp -A. Debug: kubectl exec -n {event['namespace']} <pod> -- curl <target>",
                        priority=Priority.WARNING,
                        resource_type=event.get("object_kind", "pod").lower(),
                        resource_name=event["object_name"],
                        namespace=event["namespace"]
                    ))
        except Exception as e:
            # Silently fail for network policy scanning - not all clusters have it
            pass

        return issues

    async def _scan_policy_violations(self, namespace: str = None) -> List[Issue]:
        """Detect policy engine (Kyverno) violations"""
        issues = []
        # Support "all" or empty string or None for all namespaces
        if namespace == "all" or namespace == "" or namespace is None:
            ns = None
        else:
            ns = namespace or self.namespace

        try:
            events = await self.k8s_client.get_events(ns)

            # Look for Kyverno policy violations
            kyverno_events = [e for e in events if "kyverno" in e.get("message", "").lower()
                or "policy" in e.get("reason", "").lower() and "violation" in e.get("message", "").lower()]

            for event in kyverno_events[-5:]:  # Last 5 policy violations
                issues.append(Issue(
                    id=generate_issue_id(event["object_name"], "policy-violation"),
                    title=f"Policy violation: {event['reason']}",
                    description=f"{event['message']}. Check Kyverno policies: kubectl get policyreport -n {event['namespace']}. View violations: kubectl get policyreport -n {event['namespace']} -o yaml",
                    priority=Priority.WARNING,
                    resource_type=event.get("object_kind", "pod").lower(),
                    resource_name=event["object_name"],
                    namespace=event["namespace"]
                ))
        except Exception as e:
            # Silently fail for policy violation scanning - not all clusters have it
            pass

        return issues
